rollup/l1/tracker.go

```diff
@@ -120,102 +75,255 @@ func (t *Tracker) Start() {
 	}()
 }
 
-func (t *Tracker) Stop() {
-	log.Info("stopping Tracker")
-	t.cancel()
-	t.scope.Close()
-	log.Info("Tracker stopped")
-}
+func (t *Tracker) syncLatestHead() error {
+	newHeader, err := t.client.HeaderByNumber(t.ctx, big.NewInt(int64(rpc.LatestBlockNumber)))
+	if err != nil {
+		return errors.Wrapf(err, "failed to get latest header")
+	}
+	if !newHeader.Number.IsUint64() {
+		return fmt.Errorf("received unexpected block number in L1Client: %v", newHeader.Number)
+	}
+
+	storedHeader, exists := t.headers.Get(newHeader.Number.Uint64())
+	if exists {
+		// We already processed the header, nothing to do.
+		if storedHeader.Hash() == newHeader.Hash() {
+			return nil
+		}
+
+		// Since we already processed a header at this height with different hash this means a L1 reorg happened.
+		// TODO: reset cache.
+
+		// Notify all subscribers to new LatestChainHead at their respective confirmation depth.
+		err = t.notifyLatest(newHeader, true)
+		if err != nil {
+			return errors.Wrapf(err, "failed to notify subscribers of new latest header")
+		}
+
+		return nil
+	}
+
+	// Notify all subscribers to new LatestChainHead at their respective confirmation depth.
+	err = t.notifyLatest(newHeader, false)
+	if err != nil {
+		return errors.Wrapf(err, "failed to notify subscribers of new latest header")
+	}
 
-type headerSubList struct {
-	mu   sync.Mutex
-	list []*headerSub
+	return nil
 }
 
-func (l *headerSubList) add(sub *headerSub) {
-	l.mu.Lock()
-	l.list = append(l.list, sub)
-	sort.Slice(l.list, func(i, j int) bool {
-		return l.list[i].depth < l.list[j].depth
-	})
-	l.mu.Unlock()
+func (t *Tracker) notifyLatest(newHeader *types.Header, reorg bool) error {
+	// TODO: add mutex for headers, lastSafeHeader, lastFinalizedHeader
+	t.headers.Set(newHeader.Number.Uint64(), newHeader)
+
+	t.mu.RLock()
+	defer t.mu.RUnlock()
+
+	for _, sub := range t.subscriptions[LatestChainHead] {
+		// Ignore subscriptions with deeper ConfirmationRule than the new block.
+		if newHeader.Number.Uint64() < uint64(sub.confirmationRule) {
+			continue
+		}
+
+		// 1 confirmation == latest block
+		// 2 confirmations == latest block - 1
+		// ...
+		// n confirmations == latest block - (n-1)
+		depth := uint64(sub.confirmationRule - 1)
+		headerToNotifyNumber := newHeader.Number.Uint64() - depth
+		headerToNotify, exists := t.headers.Get(headerToNotifyNumber)
+		if !exists {
+			return fmt.Errorf("header %d not found for confirmationRule %d", headerToNotifyNumber, sub.confirmationRule)
+		}
+
+		if reorg && sub.lastSentHeader != nil {
+			// The subscriber is subscribed to a deeper ConfirmationRule than the reorg depth -> this reorg doesn't affect the subscriber.
+			// Since the subscribers are sorted by ConfirmationRule, we can return here.
+			if sub.lastSentHeader.Number.Uint64() < headerToNotify.Number.Uint64() {
```

Comment:
reorg doesn't affect subscriber but new header still should be sent I think

Reply:
I don't understand what you mean here. 

Let's assume the reorg depth is 3 but the subscriber is only interested in 5 confirmations and the tip of L1 is at 103. 
Then this reorg at block 100 doesn't affect the subscriber nor do we need to notify again for block 95 as we already did that when we first processed block 100 (the one that got reorged).

The test should cover all those scenarios in detail as it is indeed quite tricky

Reply:
Let's assume that we called `notifyLatest` with `newHeader` on height `103` and `reorg` set to `true`. Then for subscriber with depth 5 `headerToNotifyNumber` will be 98. And even if last send number for that subscriber was less then 98 and reorg is true, we can inform subscriber that there is a new block on depth 5

Reply:
You are right. I'll need to do the reorg implementation differently.